cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.proto;
+
+import javax.annotation.Nullable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Represent preferences
+ */
+public class PreferencesDetail {
+  private final Map<String, String> properties;
+  /**
+   * Sequence id of operations on the preferences
+   */
+  private final Long seqId;
+  /**
+   * Whether it is a resolved preferences or not.
+   */
+  private boolean resolved;
+
+  public static PreferencesDetail merge(PreferencesDetail left, PreferencesDetail right) {
+    if (left == null && right == null) {
```

Comment:
by convention, anything that isn't annotated as `@Nullable` is assumed to be non-null, and in general we try to avoid null values when possible. When would be trying to merge a null?

Reply:
We don't such case of resolve(null, non-null), but added code to protect against that. 
Removed.